src/write/encoder_string_writer.rs

```diff
@@ -0,0 +1,108 @@
+use crate::Config;
+use std::io;
+use std::io::Write;
+use super::encoder::EncoderWriter;
+
+/// A `Write` implementation that base64-encodes data using the provided config and accumulates the
+/// resulting base64 in memory, which is then exposed as a String via `into_inner()`.
+///
+/// # Examples
+///
+/// ```
+/// use std::io::Write;
+///
+/// let mut enc = base64::write::EncoderStringWriter::new(base64::STANDARD);
+///
+/// enc.write_all(b"asdf").unwrap();
+///
+/// // get the resulting String
+/// let b64_string = enc.into_inner();
+///
+/// assert_eq!("YXNkZg==", &b64_string);
+/// ```
+///
+/// # Panics
+///
+/// Calling `write()` (or related methods) or `finish()` after `finish()` has completed without
+/// error is invalid and will panic.
+///
+/// # Performance
+///
+/// B64-encoded data is buffered in the heap since the point is to collect it in a String.
+pub struct EncoderStringWriter {
+    encoder: EncoderWriter<Vec<u8>>,
+}
+
+impl EncoderStringWriter {
+    /// Create a new EncoderStringWriter that will encode with the provided config.
+    pub fn new(config: Config) -> EncoderStringWriter {
+        EncoderStringWriter::from(String::new(), config)
+    }
+
+    /// Create a new EncoderStringWriter that will append to the provided string.
+    pub fn from(s: String, config: Config) -> EncoderStringWriter {
+        EncoderStringWriter { encoder: EncoderWriter::new(s.into_bytes(), config) }
+    }
+
+    /// Encode all remaining buffered data, including any trailing incomplete input triples and
+    /// associated padding.
+    ///
+    /// Once this succeeds, no further writes or calls to this method are allowed.
+    ///
+    /// Returns the base64-encoded form of the accumulated written data.
+    pub fn into_inner(mut self) -> String {
+        let buf = self.encoder.finish()
+            .expect("Writing to a Vec<u8> should never fail");
+
+        String::from_utf8(buf).expect("Base64 should always be valid UTF-8")
```

Comment:
It would be nice to accumulate directly into a `String`, to avoid first writing to a `Vec<u8>` and then copying into a `String`.

Reply:
See https://github.com/marshallpierce/rust-base64/pull/142#discussion_r494625657 for the previous discussion.

Reply:
To be clear converting `Vec -> String` does not require any copying (it is a free conversion) however the function used first checks that the bytes are valid utf8 to upload the `String`'s invariant.

Reply:
Ah, I misread and didn't catch that this was taking ownership of the Vec and not copying the memory. Nevermind, then.

I'm not concerned about paying the cost for the UTF-8 validity checking; it'd be nice to avoid that, but not worth adding unsafe code for.